test/lifecycle_module_test.dart

```diff
@@ -174,218 +254,468 @@ void main() {
-      expect(module.eventList, equals([]));
-    });
-
-    test('should warn when module is suspended if it was never loaded',
-        () async {
-      await module.suspend();
-      expect(lastLogMessage, isNotNull);
-      expect(lastLogMessage.level, equals(Level.WARNING));
-    });
-
-    test(
-        'should trigger resume events and call onResume when module is resumed',
-        () async {
-      await module.load();
-      await module.suspend();
-      module.eventList.clear();
-      await module.resume();
-      expect(module.eventList, equals(['willResume', 'onResume', 'didResume']));
-    });
-
-    test('should set isSuspended when module is resumed', () async {
-      await module.load();
-      await module.suspend();
-      expect(module.isSuspended, isTrue);
-      await module.resume();
-      expect(module.isSuspended, isFalse);
-    });
+    void testInvalidTransitions(
+        LifecycleState state, List<LifecycleState> invalidStates) {
+      invalidStates.forEach((fromState) {
+        test('should throw StateError when state is $fromState', () async {
+          await gotoState(module, fromState);
+          expectInLifecycleState(module, fromState);
+          expect(executeStateTransition(module, state), throwsStateError);
+        });
+      });
+    }
 
-    test('should do nothing when module is resumed if it was not suspended',
-        () async {
-      await module.load();
-      module.eventList.clear();
-      await module.resume();
-      expect(module.eventList, equals([]));
-    });
+    group('load', () {
+      var expectedLoadEvents = ['willLoad', 'onLoad', 'didLoad'];
+
+      test('should trigger loading events and call onLoad', () async {
+        await module.load();
+        expect(module.eventList, equals(expectedLoadEvents));
+      });
+
+      test('should set isLoading', () async {
+        expect(module.isLoading, isFalse);
+        var future = module.load();
+        expect(module.isLoading, isTrue);
+        await future;
+        expect(module.isLoading, isFalse);
+      });
+
+      test('should set isLoaded', () async {
+        expect(module.isLoaded, isFalse);
+        await module.load();
+        expect(module.isLoaded, isTrue);
+      });
+
+      test('should update module state', () async {
+        expectInLifecycleState(module, LifecycleState.instantiated);
+        var future = module.load();
+        expectInLifecycleState(module, LifecycleState.loading);
+        await future;
+        expectInLifecycleState(module, LifecycleState.loaded);
+      });
+
+      test('should return a useful future if it is already loading', () async {
+        await gotoState(module, LifecycleState.loading);
+        expect(module.isLoading, isTrue);
+        var future = module.load();
+        expect(module.isLoaded, isFalse);
+        await future;
+        expect(module.isLoaded, isTrue);
+      });
+
+      test('should return pending future if it is already loading', () async {
```

Comment:
This test and the one above are essentially duplicates. Fwiw, I like this one better,